DRA: seamless driver upgrades

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

When doing a rolling upgrade of a DRA driver daemonset, we want to replace the old pod with the new one such that:
- ResourceSlices remain unchanged (in particular, don't get removed and re-created).
- If the new instance fails, the kubelet falls back to the old instance.

This avoids disruptions and reduces overhead in the cluster.

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubernetes/issues/128964

#### Special notes for your reviewer:

We could split this up so that each commit gets merged separately, in the order in which they appear here.

#### Does this PR introduce a user-facing change?

```release-note
kubelet + DRA: For DRA driver plugins (and only for those!), the kubelet now supports a rolling update with `maxSurge > 0` in the driver's DaemonSet. A DRA driver must support this, which can be done via the k8s.io/dynamic-resource-allocation/kubeletplugin helper package.
```

/wg device-management

pohly commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -16,6 +16,56 @@ there.
+should end with a DNS domain that is unique for the plugin.
+
+If a plugin supports seamless upgrades, each instance should use a unique
+filename. Otherwise the following could happen:
```
cc @bart0sh @linxiulei @ffromani @SergeyKanzhelev 

This is my conclusion from the discussions in https://github.com/kubernetes/kubernetes/pull/127821. Agreed?

Does the rest of the documentation update match your understanding of the current code base?

pohly commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -16,6 +16,56 @@ there.
+- New instance starts, gets registered and replaces the old one.
+- Old instance stops, removing its socket.
+- The kubelet notices that, unregisters the plugin.
+- The plugin handler removes *the new* instance -> boom!
```
Describing such a version skew scenario is what we would normally describe in a KEP. But I suppose no-one has the time and motivation for writing one for the "kubelet plugin registry" feature?

pohly replied:
I tested it and it did indeed not work because there was no registered driver left after the old driver stopped and got unregistered by the kubelet. Oh how I hate to be right... :cry: 

Subset of the kubelet log output:
```
I0127 14:53:09.192006  173264 registration.go:154] "Register new DRA plugin" logger="DRA registration handler" pluginName="dra-8345.k8s.io" endpoint="/var/lib/kubelet/plugins/dra-8345.k8s.io-old-dra.sock"
I0127 14:53:12.265032  173264 registration.go:154] "Register new DRA plugin" logger="DRA registration handler" pluginName="dra-8345.k8s.io" endpoint="/var/lib/kubelet/plugins/dra-8345.k8s.io-new-dra.sock"
I0127 14:53:16.671448  173264 registration.go:226] "Deregister DRA plugin" logger="DRA registration handler" pluginName="dra-8345.k8s.io" endpoint="/var/lib/kubelet/plugins/dra-8345.k8s.io-new-dra.sock"
```

pohly replied:
Note that the `endpoint="/var/lib/kubelet/plugins/dra-8345.k8s.io-new-dra.sock"` in that last log entry comes from the DRA plugin manager. It's not getting passed in by the plugin watcher, which knows that `/var/lib/kubelet/plugins/dra-8345.k8s.io-old-dra.sock` got removed.

pohly replied:
This rollout problem could be solved with node labeling: nodes where the kubelet support a rolling update have to be labeled with a certain label, let's say "kubernetes.io/kubelet-plugin-seamless-update".

Then a plugin gets deployed with two DaemonSet variants, one for nodes without that label (traditional style), the other for nodes with it (`maxSurge > 0`, different sockets). We might need pod anti-affinity to prevent running pods from the different sets on the same node when the label gets added or removed. But pod-anti-affinity is expensive, so we have to think hard whether it's really needed. SIG Scheduling...

But who maintains that label? The admin? The kubelet itself? NFD? SIG Node...

This screams "a KEP is needed".

pohly replied:
Added to the README.md, without specifically naming this special label or how it gets set.

One option is that we decide that the changes in this PR are small enough to be merged without a KEP. They only affect DRA which is still only a beta API in early adoption. We know how to reach out to DRA authors and folks setting up clusters with support for it. IMHO this is something that could just be done in 1.33 and treated as an incremental improvement.

The other option is the full KEP route, with support for Device Plugin, CSI, and DRA. This can include standardizing an official node label and a feature gate to control any functional changes. But someone needs to step up to do this.

I'd like to hear from SIG leads about this.

cc @SergeyKanzhelev @haircommander @ffromani

pohly committed:
kubelet: document seamless upgrade support and guidance

This tries to capture the current state of affairs and a potential plan for
supporting seamless upgrades better.

pohly committed:
DRA kubelet: give DRA drivers a 30 second grace period for updates

When doing an update of a DaemonSet, first the old pod gets stopped and
then the new one is started. This causes the kubelet to remove all
ResourceSlices directly after removal and forces the new pod to recreate all of
them.

Now the kubelet waits 30 seconds before it deletes ResourceSlices. If a new
driver registers during that period, nothing is done at all. The new driver
finds the existing ResourceSlices and only needs to update them if something
changed.

The downside is that if the driver gets removed permanently, this creates a
delay where pods might still get scheduled to the node although the driver is
not going to run there anymore and thus the pods will be stuck.

pohly commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -16,6 +16,56 @@ there.
+`<plugin name>[-<some optional string>].sock` as filename. `<plugin name>`
+should end with a DNS domain that is unique for the plugin.
+
+If a plugin supports seamless upgrades, each instance should use a unique
```
I left it open whether rolling upgrades work when using the same filename. I ran an experiment and it **failed**. The failure scenario is similar to https://github.com/kubernetes/kubernetes/pull/129832/files#r1930224972:

```
# Old driver is registered.
I0127 15:17:35.543221  205803 registration.go:154] "Register new DRA plugin" logger="DRA registration handler" pluginName="dra-7204.k8s.io" endpoint="/var/lib/kubelet/plugins/dra-7204.k8s.io-dra.sock"
# New driver unlinks the socket, causing de-registration.
I0127 15:17:37.014042  205803 registration.go:226] "Deregister DRA plugin" logger="DRA registration handler" pluginName="dra-7204.k8s.io" endpoint="/var/lib/kubelet/plugins/dra-7204.k8s.io-dra.sock"
# New driver recreates the socket, gets registered.
I0127 15:17:38.613825  205803 registration.go:154] "Register new DRA plugin" logger="DRA registration handler" pluginName="dra-7204.k8s.io" endpoint="/var/lib/kubelet/plugins/dra-7204.k8s.io-dra.sock"
# Old driver shuts down cleanly, removing the socket, causing de-registration.
I0127 15:17:40.017121  205803 registration.go:226] "Deregister DRA plugin" logger="DRA registration handler" pluginName="dra-7204.k8s.io" endpoint="/var/lib/kubelet/plugins/dra-7204.k8s.io-dra.sock"
```

But wait! In my experiment I used a ReplicaSet to control the timing of starting and stopping pods. I could have sworn someone said that a DaemonSet supports running pods in parallel on a node for a smooth upgrade path. But looking at https://kubernetes.io/docs/tasks/manage-daemon/update-daemon-set/#daemonset-update-strategy it explicitly says "At most one pod of the DaemonSet will be running on each node during the whole update process."

All of the considerations for "what if two pods run in parallel" become pretty much pointless if there is no way to ever get into that state.

@linxiulei: how did you intend https://github.com/kubernetes/kubernetes/pull/127821 to be used in practice when there's never a situation where one pod can take over seamlessly from another?

linxiulei replied:
> it explicitly says "At most one pod of the DaemonSet will be running on each node during the whole update process."

I also noticed it but  [doc](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/daemon-set-v1/#DaemonSetSpec) of `updateStrategy.rollingUpdate.maxSurge` says something conflicting 

> The maximum number of nodes with an existing available DaemonSet pod that can have an updated DaemonSet pod during during an update.

linxiulei replied:
Just FTR, I tested:

1. deploy a DS with `updateStrategy.rollingUpdate.maxSurge=10` 
2. run `kubectl  rollout restart ds myDS`
3. check kubelet logs in any nodes and observed the new DS pod was Added, became ready, and then old DS pod was removed.

pohly replied:
Okay, so `maxSurge` is the key: by default, the statement from https://kubernetes.io/docs/tasks/manage-daemon/update-daemon-set/#daemonset-update-strategy ("At most one pod of the DaemonSet will be running on each node during the whole update process.") is correct. It just doesn't mention anywhere on that page that this default can be changed.

So without an enhanced kubelet, `maxSurge` must be zero (problem above). Will add to the doc.

With an enhanced kubelet, `maxSurge` can be non-zero as long as the socket paths are different.

Everything else is broken, leading to the rollout problem I mentioned in https://github.com/kubernetes/kubernetes/pull/129832#discussion_r1930224972.

pohly replied:
As there is a way to do it with a DaemonSet, I went ahead and implemented it for DRA. I also updated the README.md to explain how it can be used.

@linxiulei: we could need to merge this doc update first, then your PR could update with "and now Device Plugin supports it", similar to how I did it for DRA in the last commit of this PR.

@bart0sh: please also have a look.

linxiulei replied:
> So without an enhanced kubelet, maxSurge must be zero (problem above). Will add to the doc.

I still think there is a way to achieve safe semi-HA with current implementation of DS:

1. maxSurge is non-zero and so new instance is started while the old is running
2. although new instance pod is running and ready, its implementation should be able to detect where is old one running as the socket exists so that the new socket is not created
3. DS controller notices the new instance pod is ready, so deleting the old one
4. kubelet deregisters the old one
5. this step is tricky, the new instance notices the socket is gone, and creates socket for registration
6. kubelet registers the new instance

This is semi-HA, because there is still a very short period the device resource is deregistered. Also, step 5 may have race condition that kubelet registers the new instance before kubelet completes deregisteration of old one as the deletion of the socket is not a solid signal to mark the deregistration of old instance. However, a slightly more complex implementation in device-plugin could just watch the node object for the resource status to mitigate this race condition.

pohly replied:
I agree, this can work. But it seems pretty fragile to me. "Old socket exists" is not enough, the new driver also needs to detect whether the old driver is still listening. The old pod gets killed with SIGKILL, doesn't remove the socket, and the new pod never takes over because the socket exists.

It doesn't solve the seamless handover. The new pod has to remove the socket, so the kubelet will deregister the DRA driver and delete ResourceSlices unless we do the complicated "wait for a certain period of time" dance. I don't think that this is worth it.

pohly commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -98,7 +98,10 @@ func NewManagerImpl(kubeClient clientset.Interface, stateFileDirectory string, n
-	return cache.PluginHandler(dra.NewRegistrationHandler(m.kubeClient, m.getNode))
+	// The time that DRA drivers have to come back after being unregistered
+	// before the kubelet removes their ResourceSlices.
+	wipingDelay := 5 * time.Minute
```
After I [finally understood](https://github.com/kubernetes/kubernetes/pull/129832#discussion_r1930628588) that a seamless upgrade of a DaemonSet always includes some downtime, I focused on minimizing the impact.

Now ResourceSlices are preserved if the upgrade finishes within 5 minutes. The downside is that it takes 5 minutes longer in case of a permanent driver uninstall to prevent scheduling pods which depend on the driver.

Which one is more important?

cc @klueska

pohly replied:
> After I https://github.com/kubernetes/kubernetes/pull/129832#discussion_r1930628588 that a seamless upgrade of a DaemonSet always includes some downtime, I focused on minimizing the impact.

Wrong - if `maxSurge > 0`, two pods may run at the same time.

This change here may still be useful for the traditional deployment with `maxSurge = 0` (the default).

bart0sh replied:
Why it's 5 minutes? Can it be shorter?

pohly replied:
5 minutes is what most E2E tests use as "time to wait for pod start". We can make it longer or shorter.

bart0sh replied:
I'd vote for making it shorter to detect permanent plugin failures faster. How about 3 min?

pohly replied:
Added the pros and cons for choosing the duration in the comment for it.

pohly replied:
We can make it shorter. But personally I am leaning towards not doing this at all because all choices have drawbacks. Instead, DRA drivers should use seamless upgrade if they care about not dropping ResourceSlices during an upgrade, which of course hinges on actually supporting it.

bart0sh replied:
I was mostly concerned with permanent plugin failures, not with upgrades.
If 5 minutes is what we use here, I should probably increase delays/intervals in https://github.com/kubernetes/kubernetes/pull/129799 as well. WDYT?

pohly replied:
That seems different. Here we need a delay that covers the duration between "one pod stops normally, other pod starts". The kubelet detects the pod exit because the socket gets removed.

Your PR is about scenarios where a running pod is just unresponsive (which shouldn't last 5 minutes) vs. gone forever after an unclean shutdown. Anyway, explain in your PR how the constants were chosen and what the pros and cons of making them longer or shorter are and we can discuss there.

pohly replied:
> I was mostly concerned with permanent plugin failures, not with upgrades.

But we need to consider both. What is good for one (short delay after failure) is bad for the other (not enough time to finish the upgrade, which may include image pulling).

johnbelamaric replied:
> We can make it shorter. But personally I am leaning towards not doing this at all because all choices have drawbacks. Instead, DRA drivers should use seamless upgrade if they care about not dropping ResourceSlices during an upgrade, which of course hinges on actually supporting it.

Does "not doing this at all" mean unregistering results in immediate deletion? Would that happen during some sort of intermittent failure, or just if there was an orderly destruction and recreation? I think a bit of tolerance may be helpful. I agree with the general statement that if they want seamless upgrade, they should use `maxSurge > 0`. But if a user, say, deletes the pod, will it unregister and then reregister on recreation? Do we want resource slices to go away in that case?

pohly replied:
> But if a user, say, deletes the pod, will it unregister and then reregister on recreation?

Yes. Slices then get deleted on unregistration, immediately. That is the current behavior, the rationale being that we want to stop scheduling pods as soon as possible.

> I think a bit of tolerance may be helpful.

How large would you make that tolerance?

johnbelamaric replied:
I was thinking something short, just to prevent flapping during intermittent failures or crashes. Would 30s make a difference? A minute?

pohly replied:
For a container restart, 30 second might help. I'm not sure how much it would help for a real driver update, but that was the case that I would like to see handled via `maxSurge` anyway. Let's go for 30 seconds.

pohly replied:
I switched to 30 seconds.

pohly committed:
DRA kubelet: support rolling upgrades

The key difference is that the kubelet must remember all plugin instances
because it could always happen that the new instance dies and leaves only the
old one running.

The endpoints of each instance must be different. Registering a plugin with the
same endpoint as some other instance is not supported and triggers an error,
which should get reported as "not registered" to the plugin. This should only
happen when the kubelet missed some unregistration event and re-registers the
same instance again. The recovery in this case is for the plugin to shut down,
remove its socket, which should get observed by kubelet, and then try again
after a restart.

pohly commented on pkg/kubelet/cm/dra/plugin/registration.go:
```diff
@@ -209,9 +209,10 @@ func (h *RegistrationHandler) RegisterPlugin(pluginName string, endpoint string,
+	if err := draPlugins.add(pluginInstance); err != nil {
+		cancel(err)
+		// No wrapping, the error already contains details.
+		return err
```
From the commit message:

> The endpoints of each instance must be different. Registering a plugin with the same endpoint as some other instance is not supported and triggers an error, which should get reported as "not registered" to the plugin. This should only happen when the kubelet missed some unregistration event and re-registers the same instance again. The recovery in this case is for the plugin to shut down, remove its socket, which should get observed by kubelet, and then try again after a restart.

I couldn't think of a scenario where registering the same endpoint again without first unregistering it is okay.

@linxiulei @bart0sh: agreed?

bart0sh replied:
Should it be handled by the pluginwatcher, so it works for all types of Kubelet plugins?

pohly replied:
Maybe? In this PR I wanted to avoid making sweeping changes that potentially break other plugin types.

bart0sh commented on pkg/kubelet/cm/dra/plugin/plugins_store.go:
```diff
@@ -37,43 +40,63 @@ func (s *pluginsStore) get(pluginName string) *Plugin {
-	if replacedPlugin != nil && replacedPlugin.cancel != nil {
-		replacedPlugin.cancel(errors.New("plugin got replaced"))
+	for _, oldP := range s.store[p.name] {
+		if oldP.endpoint == p.endpoint {
```
Can you describe a scenario when two plugins with the same endpoint end up here?

pohly replied:
No, that's why I asked in https://github.com/kubernetes/kubernetes/pull/129832#discussion_r1932612281 about this.

bart0sh commented on pkg/kubelet/cm/dra/plugin/plugins_store.go:
```diff
@@ -37,43 +40,63 @@ func (s *pluginsStore) get(pluginName string) *Plugin {
+	var p *Plugin
+	last := false
+	var i int
+	for i, p = range instances {
+		if p.endpoint == endpoint {
+			break
+		}
+	}
+	if i == len(instances) {
+		return nil, false
+	}
```
Would it make sense to use [slices.IndexFunc](https://pkg.go.dev/slices#IndexFunc) here?

pohly replied:
Yes, might be shorter. I'll try it.

pohly replied:
Changed.

bart0sh commented on pkg/kubelet/cm/dra/plugin/plugins_store.go:
```diff
@@ -37,43 +40,63 @@ func (s *pluginsStore) get(pluginName string) *Plugin {
+		last = true
+		delete(s.store, pluginName)
+	} else {
+		s.store[pluginName] = slices.Delete(instances, i, i+1)
```
Is it possible that slice is empty? Would slices.Delete panic?

pohly replied:
Empty slice implies that the plugin wasn't found, which is handled above (`if i == len(instances) { return ... }`). Here `i` is a valid index in a non-empty slice.

bart0sh commented on pkg/kubelet/cm/dra/plugin/plugins_store.go:
```diff
@@ -37,43 +40,58 @@ func (s *pluginsStore) get(pluginName string) *Plugin {
-	delete(s.store, pluginName)
-
-	return p
+	return p, last
```
Nit:
```suggestion
	return p, len(instances) == 0
```

pohly replied:
That repeats the earlier if check and actually is wrong. It would have to be `len(instances) == 1`, as above.

pohly replied:
That saves one line and increases clarify - changed.

bart0sh replied:
Perhaps using [slices.DeleteFunc](https://pkg.go.dev/slices#DeleteFunc) would be even more clear.

pohly replied:
I need the deleted entry. `slices.DeleteFunc` doesn't give me that.

bart0sh commented on pkg/kubelet/pluginmanager/plugin_manager_test.go:
```diff
@@ -157,7 +173,12 @@ func TestPluginRegistration(t *testing.T) {
+		// And unregister.
+		fakeHandler.Reset()
+		require.NoError(t, p.Stop())
+		waitForDeRegistration(t, fakeHandler, pluginName, socketPath)
```
As it tests also for de-registration now, perhaps test name should be changed as well.

pohly replied:
Changed to `TestPluginManager`.

bart0sh commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -16,6 +16,87 @@ there.
+filenames is *not* compatible with a kubelet version that does not have support
+for seamless upgrades yet. It breaks like this:
+
+- New instance starts, gets registered and replaces the old one.
```
How is it possible? An attempt to listen on the same socket would fail for the new instance. If new instance removes the socket, then old one will be unregistered by the pluginwatcher.

pohly replied:
This section is specifically about "seamless upgrades and per-instance socket filenames". The sockets are therefore not the same.

nojnhuh commented on pkg/kubelet/cm/dra/plugin/registration.go:
```diff
@@ -179,9 +209,19 @@ func (h *RegistrationHandler) RegisterPlugin(pluginName string, endpoint string,
+	defer h.mutex.Unlock()
+	if cancel := h.pendingWiping[pluginName]; cancel != nil {
+		(*cancel)(errors.New("new plugin instance registered"))
+		h.pendingWiping[pluginName] = nil
```
I noticed it looks like we never call `delete()` on the `pendingWiping` map, so the map might grow forever even if it's filled with `nil` values. As long as the second `ok`/`exists` value of a map lookup isn't used, would this be equivalent to here and the other place where the value in the map is set to `nil`?

```suggestion
		delete(h.pendingWiping, pluginName)
```

I can't imagine that would cause any real problems by itself unless millions of different DRA plugins are registered over the lifetime of this kubelet though.

pohly replied:
I need to look into the details, but it is indeed better to delete obsolete entries.

pohly commented on test/e2e/dra/dra.go:
```diff
@@ -1399,6 +1399,126 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
+		if updateDuration <= 3*time.Minute {
+			newSlices, err := listSlices(ctx)
+			framework.ExpectNoError(err, "list slices again")
+			gomega.Expect(newSlices.Items).To(gomega.ConsistOf(oldSlices.Items))
```
`gomega.ConsistOf` uses `gomega.Equal` to compare the items, which includes comparing the UIDs. The only thing that may differ is the order.

pohly commented on test/e2e/dra/dra.go:
```diff
@@ -1399,6 +1399,126 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
+		b.create(ctx, claim, pod)
+		b.testPod(ctx, f.ClientSet, pod)
+
+		// The slices should have survived the update, but only if it happened
+		// quickly enough. If it took too long, the kubelet considered the driver
+		// gone and removed them.
+		if updateDuration <= 3*time.Minute {
```
I'm wary about E2E flakes here. Making assumptions about how quickly things are done tends to fail occasionally.

pohly commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -16,6 +16,87 @@ there.
+- The plugin handler removes *the new* instance because it ignores the socket path -> no instance left.
+
+Plugin authors either have to assume that the cluster has a recent enough
+kubelet or rely on labeling nodes with support. Then the plugin can use one
```
Unresolved thread about node labeling (attached to modified line and thus "outdated" according to GitHub): https://github.com/kubernetes/kubernetes/pull/129832#discussion_r1930224972

bart0sh commented:
/triage accepted

pohly commented:
/milestone v1.33

tico88612 commented:
Hi folks,
We'd like to remind you of the code freeze time. If there's anything we can do, please let us know.
The code freeze is starting 02:00 UTC Friday 21st March 2025 (about 5 weeks from now). Please make sure the PR has both lgtm and approved labels before the code freeze. Thanks!

johnbelamaric commented:
/assign @SergeyKanzhelev

pohly commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -16,6 +16,87 @@ there.
+back to the old instance because that old instance is not listening to the
+socket that is available under `plugin.example.com-reg.sock`.
+
+This can be achieved in a DaemonSet by passing the UID of the pod into the pod
+through the downward API. New instances may try to clean up stale sockets of
+older instances, but have to be absolutely sure that those sockets really
+aren't in use anymore. Each instance should catch termination signals and clean
+up after itself. Then sockets only leak during abnormal events (power loss,
+killing with SIGKILL).
```
I pushed an update which temporarily merges https://github.com/kubernetes/kubernetes/pull/130700 and then adds further changes in the helper.

As far as I am concerned, "seamless upgrade" support is now as complete as we need it to be for 1.33. What is missing is finding stale socket files from previous pods. I have an idea how to do that, but don't see it as essential.

/hold cancel

pohly committed:
DRA kubeletplugin: add RollingUpdate

When the new RollingUpdate option is used, the DRA driver gets deployed such
that it uses unique socket paths and uses file locking to serialize gRPC
calls. This enables the kubelet to pick arbitrarily between two concurrently
instances. The handover is seamless (no downtime, no removal of ResourceSlices
by the kubelet).

For file locking, the fileutils package from etcd is used because that was
already a Kubernetes dependency. Unfortunately that package brings in some
additional indirect dependency for DRA drivers (zap, multierr), but those
seem acceptable.

bart0sh commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -98,7 +98,20 @@ func NewManagerImpl(kubeClient clientset.Interface, stateFileDirectory string, n
+	// 30 seconds might be long enough for a simple container restart.
+	// If a DRA driver wants to be sure that slices don't get wiped,
+	// it should use rolling updates.
+	wipingDelay := 30 * time.Second
```
Would it make sense to make `wipingDelay` a constant?

pohly replied:
Can be done, but isn't used elsewhere, so it wouldn't make much of a difference.

bart0sh commented on pkg/kubelet/cm/dra/plugin/registration.go:
```diff
@@ -220,16 +260,51 @@ func (h *RegistrationHandler) validateSupportedServices(pluginName string, suppo
+		// This logger includes endpoint and pluginName.
 		logger := klog.FromContext(p.backgroundCtx)
-		logger.V(3).Info("Deregister DRA plugin", "endpoint", p.endpoint)
+		logger.V(3).Info("Deregister DRA plugin", "lastInstance", last)
```
does this logger already have a plugin name and endpoint? If not, it would be good to add them to the message.

pohly replied:
Yes, from `RegisterPlugin`. That sets "pluginName" and "endpoint".

bart0sh commented on pkg/kubelet/cm/dra/plugin/registration.go:
```diff
@@ -220,16 +260,51 @@ func (h *RegistrationHandler) validateSupportedServices(pluginName string, suppo
+		// the plugin is canceled.
+		logger = klog.FromContext(h.backgroundCtx)
+		logger = klog.LoggerWithName(logger, "driver-cleanup")
+		logger = klog.LoggerWithValues(logger, "pluginName", pluginName)
```
do we want to add `endpoint` as well?

pohly replied:
No, I don't think so. This logger is for the cleanup operation for the entire driver and not tied to one particular endpoint. The removal of the last endpoint is only what triggers that cleanup.

bart0sh commented on pkg/kubelet/cm/dra/plugin/registration.go:
```diff
@@ -220,16 +260,51 @@ func (h *RegistrationHandler) validateSupportedServices(pluginName string, suppo
+				// is the current entry. Perhaps it already got replaced.
+				cancel(errors.New("wiping done"))
+				if h.pendingWiping[pluginName] == &cancel {
+					h.pendingWiping[pluginName] = nil
```
Would it make sense to delete `pluginName` from the map instead of setting its value to nil?

pohly replied:
Yes, changed. That prevents growing the map over time with stale plugin names.

bart0sh commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -16,6 +16,87 @@ there.
+
+- New instance starts, gets registered and replaces the old one.
+- Old instance stops, removing its socket.
+- The kubelet notices that, unregisters the plugin.
```
So, it looks possible that `PrepareResources` is called for the old instance and `UnprepareResources` - for the new one. How new instance would "know" that the resource is prepared when receiving `UnprepareResources` request?  Is this info stored in the slice and new instance should fetch it before processing the request?

pohly replied:
How to persist state across container restarts is a problem that DRA drivers must solve themselves. This PR doesn't change anything in that regard.

They can use local files for that or reconstruct the current state of the node on a container start. They cannot use anything in the API, there are no fields for that.

bart0sh replied:
> How to persist state across container restarts is a problem that DRA drivers must solve themselves. 

Is this mentioned anywhere in the documentation? DRA drivers developers should be informed about this.
BTW, this could be different from handling restarts as far as I understand. In this scenario it's possible that two drivers are running at the same time and not restarting. Kubelet will call PrepareResources for the old driver and UnprepareResources for another one and both should be able to handle this somehow.

pohly replied:
> Is this mentioned anywhere in the documentation? DRA drivers developers should be informed about this.

I'm pretty sure we mentioned it. The DRA example driver implements it.

> BTW, this could be different from handling restarts as far as I understand. In this scenario it's possible that two drivers are running at the same time and not restarting. Kubelet will call PrepareResources for the old driver and UnprepareResources for the new one and both should be able to handle this somehow.

The `RollingUpdate` calls out the "running two plugin instances in parallel" aspect. It's left to the DRA driver author to recognize what that means. Let me be more explicit...

pohly replied:
```
...
// There is no guarantee which of the two instances are used by kubelet.
// For example, it can happen that a claim gets prepared by one instance
// and then needs to be unprepared by the other. Kubelet then may fall back
// to the first one again for some other operation. In practice this means
// that each instance must be entirely stateless across method calls.
// Serialization (on by default, see [Serialize]) ensures that methods
// are serialized across all instances through file locking. The plugin
// implementation can load shared state from a file at the start
// of a call, execute and then store the updated shared state again.
...
func RollingUpgrade ...
```

tico88612 commented:
Hello, @pohly @bart0sh @nojnhuh 
Appreciate all of your efforts with this PR! Is the plan still to resolve it for v1.33 release?
If so, a gentle reminder that the code freeze has started [02:00 UTC Friday 21st March 2025](https://everytimezone.com/s/2c5e9275) . Please make sure any PRs have both lgtm and approved labels ASAP, and file an [Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md) if you haven't done it yet.
Thanks!

pohly commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -16,6 +16,87 @@ there.
+
+**Warning**: Such a seamless upgrade is only supported for DRA at the moment.
+
+### In a plugin
```
Yes, added.

pohly commented:
I added the following patch and then rebased:

```patch
diff --git a/pkg/kubelet/cm/dra/plugin/plugins_store.go b/pkg/kubelet/cm/dra/plugin/plugins_store.go
index 1f6383f93a0..cc2fc240b5b 100644
--- a/pkg/kubelet/cm/dra/plugin/plugins_store.go
+++ b/pkg/kubelet/cm/dra/plugin/plugins_store.go
@@ -69,7 +69,7 @@ func (s *pluginsStore) add(p *Plugin) error {
 	return nil
 }
 
-// Delete lets you remove one endpoint for a DRA Plugin.
+// remove lets you remove one endpoint for a DRA Plugin.
 // This method is protected by a mutex. It returns the
 // plugin if found and true if that was the last instance
 func (s *pluginsStore) remove(pluginName, endpoint string) (*Plugin, bool) {
diff --git a/pkg/kubelet/cm/dra/plugin/registration.go b/pkg/kubelet/cm/dra/plugin/registration.go
index 27221c00e11..570ce4a0579 100644
--- a/pkg/kubelet/cm/dra/plugin/registration.go
+++ b/pkg/kubelet/cm/dra/plugin/registration.go
@@ -300,7 +300,7 @@ func (h *RegistrationHandler) DeRegisterPlugin(pluginName, endpoint string) {
 				// is the current entry. Perhaps it already got replaced.
 				cancel(errors.New("wiping done"))
 				if h.pendingWiping[pluginName] == &cancel {
-					h.pendingWiping[pluginName] = nil
+					delete(h.pendingWiping, pluginName)
 				}
 			}()
 			h.wipeResourceSlices(ctx, h.wipingDelay, pluginName)
diff --git a/pkg/kubelet/pluginmanager/pluginwatcher/README.md b/pkg/kubelet/pluginmanager/pluginwatcher/README.md
index 87280fa93c7..9403829a2fb 100644
--- a/pkg/kubelet/pluginmanager/pluginwatcher/README.md
+++ b/pkg/kubelet/pluginmanager/pluginwatcher/README.md
@@ -32,6 +32,11 @@ setting `maxSurge` to a value larger than zero enables such a seamless upgrade.
 
 ### In a plugin
 
+*Note*: For DRA, the
+[k8s.io/dynamic-resource-allocation](https://pkg.go.dev/k8s.io/dynamic-resource-allocation/kubeletplugin)
+helper package offers the `RollingUpdate` option which implements the socket
+handling as described in this section.
+
 To support seamless upgrades, each plugin instance must use a unique
 socket filename. Otherwise the following could happen:
 - The old instance is registered with `plugin.example.com-reg.sock`.
```

@bart0sh @nojnhuh: please take another look.

pohly commented:
/priority important-soon

Would be good to have in 1.33.

bart0sh commented:
/lgtm

mrunalp commented on pkg/kubelet/cm/dra/plugin/registration.go:
```diff
@@ -53,6 +54,18 @@ type RegistrationHandler struct {
+	//
+	// Must use pointers to functions because the entries have to
+	// be comparable.
+	pendingWiping map[string]*context.CancelCauseFunc
```
nit: wdyt of naming this `pendingWipes`?

pohly replied:
Ack, changed.

pohly replied:
@mrunalp: [pushed](https://github.com/kubernetes/kubernetes/compare/3b0a27af309f465fcd5a3b92439fe58e57c14a17..582b421393d0fad2ad4a83feba88977ac4434662), please take another look.

bart0sh replied:
/lgtm

mrunalp commented on pkg/kubelet/cm/dra/plugin/plugins_store.go:
```diff
@@ -37,43 +40,57 @@ func (s *pluginsStore) get(pluginName string) *Plugin {
+	}
+	// Heuristic: pick the most recent one. It's most likely
+	// the newest, except when kubelet got restarted and registered
+	// all running plugins in random order.
```
Do we want to remember this info on disk to prevent random load order?

pohly replied:
I think this is so unlikely (kubelet being restarted in the middle of a driver update) and not harmful when it picks the older (we explicitly state that both instances must be usable interchangeably) that it doesn't justify making the code more complex.

TLDR: the heuristic should be sufficient.

gnufied commented:
/lgtm

gnufied commented:
/approve
